apps/wallet/src/background/Keyring.ts

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { Ed25519Keypair } from '@mysten/sui.js';
+import { BehaviorSubject } from 'rxjs';
+import Browser from 'webextension-polyfill';
+
+import { encrypt, decrypt } from '_shared/cryptography/keystore';
+import { generateMnemonic } from '_shared/utils/bip39';
+
+import type { Keypair } from '@mysten/sui.js';
+
+const STORAGE_KEY = 'vault';
+
+class Keyring {
+    private readonly _locked = new BehaviorSubject<boolean>(true);
+    private _encryptedMnemonic: Promise<string | null>;
+    private _keypair: Keypair | null = null;
+
+    constructor() {
+        this._encryptedMnemonic = this.loadMnemonic();
+    }
+
+    // Creates a new mnemonic and saves it to storage encrypted
+    public async createMnemonic(password: string) {
+        if (await this.isWalletInitialized()) {
+            throw new Error(
+                'Mnemonic already exists, creating a new one will override it. Clear the existing one first.'
+            );
+        }
+        const encryptedMnemonic = await encrypt(
+            password,
+            Buffer.from(generateMnemonic(), 'utf8')
+        );
+        await this.storeEncryptedMnemonic(encryptedMnemonic);
+        this._encryptedMnemonic = Promise.resolve(encryptedMnemonic);
+    }
+
+    public lock() {
+        this._keypair = null;
+        this._locked.next(true);
+    }
+
+    public async unlock(password: string) {
+        this._keypair = Ed25519Keypair.deriveKeypair(
```

Comment:
I think we should also keep the mnemonic itself stored on the class here so that we can derive additional keypairs when we have multi-account.

Reply:
Let's think about this at that point, we can create all the keys when we unlock and just ask the user for the password when he wants to add an new one.

Reply:
IMO account creation should be possible without the password.

Reply:
Yeah I agree, the easier the better. Since it's not necessary now we could change it when it will be necessary. Anyway I am changit it now

Reply:
in the future this can be the master keypair and we can derive additional keypairs.